### ISSUE TITLE
Some test of speed for ctranslate and huggingface based on Llama

### ISSUE THREAD

**Issue by Snowdar** [OpenNMT/CTranslate2#1276]
Here is some test of speed to analyze the stability of GPU generation between huggingface (torch) and ctranslate.

Test params
```
GPU = A100 x 1
prompt tokens = 435
generate tokens = 29
run times = 5 # continuous independent running
cuda warmup = yes # test after run first example
hf_data_type = float16
ctranslate_data_type = float16
prompt_cache = no # prompt eval is included in the generation of first token
```

 <details> <summary>Elapsed time (s) of tokens of huggingface</summary>

```
[1]
0.056...0.097...0.133...0.171...0.207...0.244
0.280...0.319...0.355...0.391...0.427...0.464
0.500...0.537...0.574...0.610...0.647...0.683
0.720...0.759...0.796...0.832...0.869...0.905
0.942...0.978...1.018...1.057...1.103

[2]
0.056...0.103...0.148...0.191...0.236...0.277
0.323...0.381...0.434...0.485...0.545...0.597
0.642...0.688...0.740...0.791...0.847...0.902
0.958...1.011...1.067...1.116...1.164...1.211
1.262...1.319...1.382...1.462...1.529
# bad performance due to the other tokens responspe lag (Is it caused by cuda stability?)

[3]
0.060...0.114...0.157...0.197...0.236...0.276
0.315...0.353...0.390...0.427...0.465...0.503
0.543...0.583...0.623...0.661...0.698...0.736
0.773...0.811...0.849...0.886...0.924...0.962
0.999...1.038...1.075...1.113...1.150

[4]
0.059...0.094...0.130...0.167...0.202...0.238
0.275...0.311...0.347...0.384...0.421...0.456
0.493...0.530...0.565...0.602...0.638...0.675
0.711...0.748...0.785...0.822...0.858...0.895
0.931...0.968...1.004...1.043...1.079

[5]
0.056...0.097...0.137...0.178...0.224...0.265
0.306...0.352...0.389...0.434...0.471...0.507
0.543...0.580...0.616...0.653...0.695...0.732
0.768...0.805...0.841...0.878...0.915...0.951
0.988...1.029...1.066...1.102...1.139
```
</details>

 <details> <summary>Elapsed time (s) of tokens of ctranslate</summary>

```
[1]
0.095...0.109...0.123...0.137...0.149...0.166
0.179...0.196...0.210...0.223...0.237...0.249
0.262...0.275...0.288...0.301...0.314...0.328
0.341...0.354...0.371...0.381...0.395...0.407
0.420...0.434...0.447...0.460...0.474

[2]
0.096...0.109...0.123...0.137...0.161...0.170
0.184...0.199...0.212...0.225...0.239...0.253
0.267...0.280...0.294...0.307...0.321...0.338
0.347...0.360...0.374...0.390...0.405...0.418
0.434...0.446...0.460...0.473...0.486

[3]
0.102...0.117...0.132...0.145...0.165...0.172
0.186...0.200...0.224...0.237...0.251...0.264
0.277...0.290...0.303...0.317...0.330...0.343
0.357...0.370...0.383...0.396...0.409...0.422
0.436...0.449...0.462...0.474...0.488

[4]
0.626...0.639...0.652...0.664...0.678...0.690
0.704...0.717...0.730...0.743...0.756...0.769
0.782...0.795...0.808...0.821...0.834...0.847
0.860...0.873...0.886...0.899...0.912...0.925
0.939...0.952...0.965...0.977...0.991
# bad performance due to the first token response lag

[5]
0.089...0.102...0.115...0.128...0.141...0.154
0.167...0.180...0.193...0.206...0.220...0.232
0.245...0.258...0.272...0.285...0.298...0.311
0.324...0.337...0.350...0.363...0.376...0.389
0.402...0.415...0.428...0.441...0.455
```
</details>

Note:
1. The elapsed time of the fisrt token standards the prompt eval and first token generation (no kv cache yet)
2. The kv cache will be used in the generation of other tokens
3. Time cost of tokenizing and while grammar of python in HF generation (Ctranslate is c++) could be ignored compared with inference

Analysis:
1. Ctranslate is always faster than HF overall (see the end of elapsed time)
2. HF is fater than ctranslate in the generation of first token (see the start of elapsed time)
3. The generation of HF is more stable for the first token (0.056 ~ 0.060s)
4. The generation of Ctranslate is more stable for the tokens from the second to the end (elapsed time: end - start)

Question: How to stable the prompt eval for ctranslate to reduce the first reponse latency? And why?

**Comment by guillaumekln** [comment]
Can you post the code you are using to run HF and CTranslate2?

**Comment by Snowdar** [comment]
This is the test of MPT, it seems to be slower than llama, especially the first token (it is already version 3.15).
```
[1]
0.172...0.188...0.205...0.222...0.238...0.254
0.270...0.286...0.303...0.319...0.335...0.351
0.367...0.383...0.400...0.416...0.432...0.448
0.464...0.480...0.497...0.513...0.529...0.545
0.562...0.577...0.594...0.610...0.626

[2]
0.646...0.661...0.679...0.698...0.716...0.733
0.749...0.767...0.788...0.802...0.820...0.836
0.854...0.871...0.888...0.904...0.921...0.937
0.961...0.978...0.995...1.011...1.028...1.045
1.061...1.078...1.095...1.111...1.129

[3]
0.242...0.258...0.275...0.292...0.308...0.325
0.342...0.359...0.376...0.392...0.409...0.426
0.442...0.458...0.475...0.512...0.512...0.524
0.541...0.557...0.583...0.599...0.620...0.641
0.660...0.677...0.694...0.712...0.730

[4]
0.142...0.157...0.173...0.189...0.205...0.221
0.238...0.253...0.270...0.285...0.301...0.317
0.333...0.349...0.365...0.381...0.397...0.413
0.429...0.445...0.461...0.477...0.493...0.509
0.525...0.541...0.557...0.573...0.589

[5]
0.135...0.150...0.166...0.182...0.198...0.214
0.230...0.246...0.262...0.277...0.293...0.309
0.325...0.341...0.357...0.372...0.388...0.404
0.420...0.436...0.451...0.467...0.483...0.499
0.515...0.531...0.547...0.563...0.579
```

**Comment by Snowdar** [comment]
> Thanks for sharing the code.
> 
> I tested on an instance with a V100 and the results are pretty stable on my end. Is there something else happening on your server when you get bad performance?
> 
> Regarding the time to the first token with LLaMa, on V100 I found that CTranslate2 is slightly faster but I would also expect a clearer win here. One difference I found is that HF is forwarding the full prompt and immediately get the first token for the output logits. On the other hand, CTranslate2 will forward the first N-1 prompt tokens at once and then start the greedy decoding loop from the last prompt token. So basically the first generated token requires 2 model execution in CTranslate2. I'm investigating.
> 
> Regarding MPT, we fixed the bad performance issue when creating the ALiBi tensor that you reported in #1259. However, the performance for this part is still not good enough, especially for longer prompts. I will further optimize this code for the next version.

Thanks for your reply. Could you share your time_list of test based on my codes? 
The case of 2 model execution to generage the first token in CTranslate2 could explain the most of runs (ctranslate 0.100s vs HF 0.060s), but I am more concerned about the stability of the first token generation, sometimes you will find that it is far greater than 0.100s. Of course, I will test it again in a complete empty server (but I do not have v100 to keep consistent with your test and in the past test, there is no other process to use GPU and the core of CPU is also very free).
Finally, could you try a longger prompt (such as num_tokens > 1000 or 1500) to observe the elapsed time of first token? In my test, the longer the prompt, the easier it is to observe the unstable phenomenon of first token generation (but always ok for the second to the end).

**Comment by Snowdar** [comment]
I have also updated the code with a TokenStreamer for HF to ensure a fair comparison.

> > Can you post the code you are using to run HF and CTranslate2?
> 
> Here is the most streamlined test code and it maybe reproduce the test results.
> 
> ```
> 
> import torch
> import time
> import ctranslate2
> from threading import Thread
> from transformers import AutoTokenizer, AutoModelForCausalLM, TextIteratorStreamer
> 
> # common
> prompt_template = """### instructions
> You are a helpful assistant and your task is to chat with the user and give a compatible response. Below is a conversation with the user.
> {}
> ### user
> Hello, I need some help, I want to ask for how to learn Python? can you give me some useful suggestions to help me?
> ### response
> """
> 
> insert = "If someone ask who are you, you should reply with these sentence: I am a helpful assistant, how can I assist with you?"
> n_insert = 13 # default is 13 to get 435 tokens
> prompt = prompt_template.format('\n'.join([insert for i in range(n_insert)]))
> print(prompt)
> 
> hf_model_path = 'pretrained/vicuna-7b-v1.1'
> hf_device = torch.device('cuda:0')
> tokenizer = AutoTokenizer.from_pretrained(hf_model_path, use_fast=False)
> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
> print(input_ids.shape)
> 
> test_type = 'hf' # hf | ctranslate
> 
> def run(gen_func):
>     print("Running {} ...".format(gen_func.__name__))
>     for index in range(1+5):
>         time_list = gen_func()
>         if index == 0:
>             # skip warmup
>             continue
>         delta_list = [time_list[0]]
>         for i in range(1, len(time_list)):
>             delta_list.append(time_list[i] - time_list[i-1])
>         for i in range(len(time_list)):
>             if i % 6 == 0:
>                 print('\n', end='')
>             print("{:.3f}({:.3f})...".format(time_list[i], delta_list[i]), end='')
>         print('\n---tokens: {}---'.format(len(time_list)))
> 
> # HF
> if test_type == 'hf':
>     print("Loading hf model ...")
>     hf_model = AutoModelForCausalLM.from_pretrained(hf_model_path, low_cpu_mem_usage=True, torch_dtype=torch.float16, device_map={"": hf_device})
>     hf_model.eval()
> 
> class TokenStreamer(TextIteratorStreamer):
>     def put(self, value):
>         if len(value.shape) > 1 and value.shape[0] > 1:
>             raise ValueError("TextStreamer only supports batch size 1")
>         elif len(value.shape) > 1:
>             value = value[0]
> 
>         if self.skip_prompt and self.next_tokens_are_prompt:
>             self.next_tokens_are_prompt = False
>             return
>         
>         self.on_finalized_text(value.tolist())
> 
>     def on_finalized_text(self, tokens):
>         for token in tokens:
>             self.text_queue.put(token, timeout=self.timeout)
> 
>     def end(self):
>         self.text_queue.put(self.stop_signal, timeout=self.timeout)
> 
> def hf_generate():
>     # to keep consistent with ctranslate for device copy
>     cuda_input_ids = input_ids.to(hf_device)
>     streamer = TokenStreamer(tokenizer, skip_prompt=True)
> 
>     generate_args = dict(
>         inputs=cuda_input_ids,
>         do_sample=True,
>         top_k=40,
>         temperature=0.1,
>         top_p=1.0,
>         use_cache=True,
>         max_new_tokens=29,
>         streamer=streamer,
>     )
>     thread = Thread(target=hf_model.generate, kwargs=generate_args)
>     start = time.time()
>     thread.start()
> 
>     time_list = []
>     for new_token in streamer:
>         time_list.append(time.time() - start)
>     #     print(new_token, end='|')
>     # print('')
>     return time_list
> 
> if test_type == 'hf':
>     run(hf_generate)
> 
> # --------------------------------- #
> 
> # Ctranslate
> if test_type == 'ctranslate':
>     print("Loading ctranslate model ...")
>     ctranslate_model_path = 'pretrained/vicuna-7b-v1.1/ctranslate'
>     model = ctranslate2.Generator(ctranslate_model_path, compute_type='float16', device='cuda', device_index=0) 
> 
> def generate_ctranslate():
>     str_tokens = tokenizer.convert_ids_to_tokens(input_ids[0])
>     start = time.time()
>     step_results = model.generate_tokens(str_tokens, sampling_topk=40, sampling_temperature=0.1, max_length=29)
>     time_list = []
>     for step_result in step_results:
>         time_list.append(time.time() - start)
>         # print(step_result.token_id, end='|')
>     # print('')
>     return time_list
> 
> if test_type == 'ctranslate':
>     run(generate_ctranslate)
> ```
> 
> By the way, if the num of prompt tokens is not too many, ctranslate could be faster than HF.

**Comment by Snowdar** [comment]
Thanks for your test.
In my tests, the variable CT2_CUDA_TRUE_FP16_GEMM made no noticeable difference.

Now, I still don't know what the problem is, but it looks like it has a lot to do with my server. Therefore, I plan to dig into the code, find the specific module that increases the time consumption and try to analyze the reason.

**Comment by Snowdar** [comment]
I have found the reason. In fact, having the ```nvidia-smi lms``` and ```nvitop``` commands open for a long time on the machine affected the test. Although the CPU and GPU seemed idle, it caused some lag. Therefore, ctranslate was as stable as hf. 

Thanks for your anasysis again! I'll close this issue now.
